source/tutorials/callpackage.md

```diff
@@ -0,0 +1,283 @@
+---
+date: 2022-09-08
+authors:
+  - Norbert Melzer
+  - Valentin Gagarin
+  - Matthias Meschede
+myst:
+  html_meta:
+    "keywords": "tutorial, callPackage, override, package, customise, parameters, nix, nixpkgs"
+---
+
+(callpackage-tutorial)=
+# Package parameters and overrides with `callPackage`
+
+Nix ships with a a special-purpose programming language for creating packages and configurations: the Nix language.
+It is used to build the Nix package collection, known as {term}`Nixpkgs`.
+
+Being purely functional, the Nix language allows declaring custom functions to abstract over common patterns.
+One of the most prominent patterns in Nixpkgs is parametrisation of package recipes.
+
+## Overview
+
+Nixpkgs is sizeable software project on it's own, with coding conventions and idioms that have emerged over the years.
+It has [established a convention](https://github.com/NixOS/nixpkgs/pull/9869) of composing parameterised packages with automatic settings through a function named [`callPackage`](https://github.com/NixOS/nixpkgs/commit/fd268b4852d39c18e604c584dd49a611dc795a9b).
+This tutorial shows how to use it and why it's beneficial.
+
+### What will you learn?
+
+- Using `callPackage` to invoke package recipes that follow Nixpkgs conventions
+- Overriding package parameters
+- Creating interdependent package sets
+
+### What do you need?
+
+- Familiarity with the [Nix language](reading-nix-language)
+- First experience with [packaging existing software](packaging-tutorial)
+
+### How long does it take?
+
+- 45 minutes
+
+## Automatic function calls
+
+Create a new file `hello.nix`, which could be a typical package recipe as found in Nixpkgs:
+A function that takes an attribute set, with attributes corresponding to derivations in the top-level package set, and returns a derivation.
+
+```{code-block} nix
+:caption: hello.nix
+{ writeShellScriptBin }:
+writeShellScriptBin "hello" ''
+  echo "Hello, world!"
+''
+```
+
+:::{dropdown} Detailed explanation
+`hello.nix` declares a function which takes as argument an attribute set with one element `writeShellScriptBin`.
+[`writeShellScriptBin`](https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-writeShellScriptBin) is a function that happens to exist in Nixpkgs, a [build helper](https://nixos.org/manual/nixpkgs/unstable/#part-builders) that returns a derivation.
+The derivation output in this case is an executable shell script with the contents `echo "hello world"` named `"hello"`.
+:::
+
+Now create a file `default.nix` with the following contents:
+
+```{code-block} nix
+:caption: default.nix
+let
+  pkgs = import <nixpkgs> { };
+in
+pkgs.callPackage ./hello.nix { }
+```
+
+Realise the derivation in `default.nix` and run the executable that is produced:
```

Comment:
I don't think the terms "realise" and "derivation" work here. The Nix file contains a Nix expression that first needs to be instantiated before it's a derivation.

```suggestion
Build `default.nix` and run the executable that is produced:
```

Reply:
We haven't really converged on a robust terminology here. In the Nix manual we distinguish between [derivation](https://hydra.nixos.org/build/253249891/download/1/manual/glossary.html#gloss-derivation) (which is rather vaguely defined) and store derivation (the thing that needs to be instantiated). Also "to build" something isn't well defined (although arguably it is the specific act of "running the `builder` executable of a derivation"), while "to realise" is (and is broader than "to build"). 

In short: I'd rather leave that as is, at least for now.